### ISSUE TITLE
Proper way to check if current user can sudo

### ISSUE THREAD

**Issue by mcornella** [sudo-project/sudo#130]
I'm trying to write a check to see if the user can run `sudo`, in a way that is portable and compatible with old versions of `sudo`, and doesn't ask the user for a password.

This is what I came up, but can this be done better? I really hate having to check for the correct wording in the error message, since that seems to be brittle.

```sh
user_can_sudo() {
  # The following command has 3 parts:
  #
  # 1. Run `sudo` with `-v`. Does the following:
  #    • with privilege: asks for a password immediately.
  #    • without privilege: exits with error code 1 and prints the message:
  #      Sorry, user <username> may not run sudo on <hostname>
  #
  # 2. Pass `-n` to `sudo` to tell it to not ask for a password. If the
  #    password is not required, the command will finish with exit code 0.
  #    If one is required, sudo will exit with error code 1 and print the
  #    message:
  #    sudo: a password is required
  #
  # 3. Check for the words "may not run sudo" in the output to really tell
  #    whether the user has privileges or not. For that we have to make sure
  #    to run `sudo` in the default locale (with `LANG=`) so that the message
  #    stays consistent regardless of the user's locale.
  #
  LANG= sudo -n -v 2>&1 | grep -q "may not run sudo"
}
```

Thanks

**Comment by Samisafool** [comment]
How about this?:

```zsh
#!/usr/bin/env zsh
read -rs '?Enter your password: (It will be invisible): ' PASSWD
echo

echo "$PASSWD" |  sudo -Sv > /dev/null 2>&1 ||
{
    echo "Invalid password! Exiting..."
    exit 1
}
```

**Comment by mcornella** [comment]
Closing in favor of asking to the [sudo-users mailing list](https://www.sudo.ws/mailman/listinfo/sudo-users).

**Comment by remram44** [comment]
With the recent problems with behavior changes (#131, affects Ubuntu 22.04) and the fact that every method not only *checks* but also *refreshes* the timestamp, I went with just checking the timestamp file myself. My code probably doesn't support every single sudo setup but should unblock you if your prompt suddenly broke or if you get thousands of emails for failed logins.

https://github.com/remram44/read-sudo-timestamp
